<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>JSDoc: Source: mSigSDKScripts/mutationalSpectrum.js</title>

    <script src="scripts/prettify/prettify.js"> </script>
    <script src="scripts/prettify/lang-css.js"> </script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify-tomorrow.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc-default.css">
</head>

<body>

<div id="main">

    <h1 class="page-title">Source: mSigSDKScripts/mutationalSpectrum.js</h1>

    



    
    <section>
        <article>
            <pre class="prettyprint source linenums"><code>import { fetchURLAndCache } from "./utils.js";

function get_sbs_trinucleotide_contexts() {
  const nucleotide_bases = ["A", "C", "G", "T"];
  const substitution_types = ["C>A", "C>G", "C>T", "T>A", "T>C", "T>G"];
  let sbs_trinucleotide_contexts = [];

  for (let base_5 of nucleotide_bases) {
    for (let substitution of substitution_types) {
      for (let base_3 of nucleotide_bases) {
        sbs_trinucleotide_contexts.push(`${base_5}[${substitution}]${base_3}`);
      }
    }
  }

  return sbs_trinucleotide_contexts;
}

function standardize_substitution(ref_allele, mut_allele) {
  /*
COSMIC signatures define mutations from a pyrimidine allele (C, T) to any
other base (C>A, C>G, C>T, T>A, T>C, T>G). If a mutation in the MAF file
is defined from a reference purine allele (A, G), then we infer the substituted
base in the complementary sequence, which would be from a pyrimidine
allele due to purines and pyrimidines complementing each other in a
double-stranded DNA.
 :param ref_allele: base in the reference genome.
:param mut_allele: base in the mutated genome
:return: substitution string from pyrimidine to any other base.
*/
  var complement_seq, purines;
  complement_seq = {
    A: "T",
    C: "G",
    T: "A",
    G: "C",
  };
  purines = ["A", "G"];

  if (purines.some((v) => ref_allele.includes(v))) {
    return `${complement_seq[ref_allele]}>${complement_seq[mut_allele]}`;
  } else {
    return `${ref_allele}>${mut_allele}`;
  }
}

function init_sbs_mutational_spectra(n_records) {
  /*
Initilizes an ordered dictionary with SBS trinucleotide context as keys and
a list of counts, one for each sample.
 :param n_records: number of samples to record in the mutational spectra matrix.
:return: a dictionary of trinucleotide context and a list of counts
initialized to zeros.
*/

  let tri_nuc_context = get_sbs_trinucleotide_contexts();

  let sbs_mutational_spectra = {};

  for (var i = 0; i &lt; tri_nuc_context.length; i++) {
    let context = tri_nuc_context[i];
    sbs_mutational_spectra[context] = 0;
  }

  return sbs_mutational_spectra;
}

function standardize_trinucleotide(trinucleotide_ref) {
  // COSMIC signatures define mutations from a pyrimidine allele (C, T) to any
  // other base (C>A, C>G, C>T, T>A, T>C, T>G). If a mutation in the MAF file
  // is defined from a purine allele (A, G), then we infer the trinucleotide
  // context in the complementary sequence, which would be from a pyrimidine
  // allele due to purines and pyrimidines complementing each other in a
  // double-stranded DNA.

  // :param trinucleotide_ref: trinucleotide sequence seen in the reference genome.
  // :return: a pyrimidine-centric trinucleotide sequence.

  let complement_seq = {
    A: "T",
    C: "G",
    T: "A",
    G: "C",
  };
  let purines = "AG";
  if (purines.includes(trinucleotide_ref[1])) {
    return `${complement_seq[trinucleotide_ref[2]]}${
      complement_seq[trinucleotide_ref[1]]
    }${complement_seq[trinucleotide_ref[0]]}`;
  } else {
    return trinucleotide_ref;
  }
}

/**
 * Converts patient mutation data into mutational spectra.
 * 
 * This function processes mutation data for each patient, computes the mutational context,
 * and aggregates the results into a mutational spectrum based on the specified grouping criterion.
 * 
 * @async
 * @function convertMatrix
 * @memberof ICGC
 * @param {Array} data - The patient mutation data to be converted. Each patient's data should be an array of mutation records.
 * @param {string} [group_by="project_code"] - The property used to group mutational spectra (e.g., "project_code").
 * @param {number} [batch_size=100] - The number of mutations to process in each batch for parallelized computations.
 * @param {string} [genome="hg19"] - The genome build to use for determining mutational context if not specified in mutation data.
 * @returns {Object} - An object containing mutational spectra for each group, where keys are group identifiers.
 * @throws {Error} - If there is an error in processing the mutation data or fetching the mutational context.
 * 
 * @example
 * // Example usage:
 * const data = [
 *   [
 *     { chromosome: "1", reference_allele: "C", tumor_seq_allele2: "T", start_position: "12345", variant_type: "SNP", project_code: "PRJ001" },
 *     // Additional mutation records...
 *   ]
 * ];
 * const spectra = await convertMatrix(data, "project_code", 50, "hg38");
 * console.log(spectra);
 */


const extractFirstNumber = (str) => {
  // Match the first occurrence of one or more digits
  const match = str.match(/\d+/);
  // Convert the match to an integer and return, or return null if no match
  return match ? parseInt(match[0], 10) : null;
};

async function convertMatrix(data, group_by="project_code",  batch_size = 100, genome = "hg19",) {
  const mutationalSpectra = {};
  group_by = group_by.toLowerCase();

  for (let patient of data) {

    patient = patient.map(row =>
      Object.fromEntries(
        Object.entries(row).map(([key, value]) => [key.toLowerCase(), value])
      )
    );

    // Move the initialization of mutationalSpectrum inside the loop
    var mutationalSpectrum = init_sbs_mutational_spectra();
    var promises = [];

    for (let i = 0; i &lt; patient.length; i++) {
      
      // if patient[i]['build'] exists, then use it to determine the genome
      // if patient[i]['build'] does not exist, then use the genome parameter
      // if genome parameter is not provided, then use hg19
      if (patient[i]['build']) {
        genome = patient[i]['build'];
      } else if (!genome) {
        genome = "hg19";
      }
      

      var chromosomeNumber = patient[i]["chromosome"];
      chromosomeNumber = extractFirstNumber(chromosomeNumber);
      var referenceAllele = patient[i]["reference_allele"];
      var mutatedTo = patient[i]["tumor_seq_allele2"];
      var position = patient[i]["start_position"];
      var variantType = patient[i]["variant_type"];

      var promise = getMutationalContext(chromosomeNumber, genome, parseInt(position))
        .then((sequence) => {
          sequence = standardize_trinucleotide(sequence);
          let fivePrime = sequence[0];
          let threePrime = sequence[2];
          let mutationType = String(
            `${fivePrime}[${standardize_substitution(
              referenceAllele,
              mutatedTo
            )}]${threePrime}`
          ).toUpperCase();

          if (
            (variantType == "SNP" ||
              variantType == "single base substitution") &amp;&amp;
            !mutationType.includes("N") &amp;&amp;
            !mutationType.includes("U")
          ) {
            mutationalSpectrum[mutationType] =
              Number(mutationalSpectrum[mutationType]) + Number(1);
          }
        })
        .catch((error) => {
          console.error(error);
        });
      promises.push(promise);

      if (i % batch_size === 0 || i === patient.length - 1) {
        await Promise.all(promises);
        promises = [];
      }
    }
    // Use the patient's project_code as the key in the mutationalSpectra object
    mutationalSpectra[patient[0][group_by]] = mutationalSpectrum;
  }

  return mutationalSpectra;
}

async function getMutationalContext(chromosomeNumber, genome, startPosition) {
  const chrName = String(chromosomeNumber);
  const startByte = startPosition - 2;
  const endByte = startPosition;

  const alternative = await (
    await fetchURLAndCache("HG19",
      `https://api.genome.ucsc.edu/getData/sequence?genome=${genome};chrom=chr${chrName};start=${startByte};end=${
        endByte + 1
      }`
    )
  ).json();

  const sequence = alternative.dna;
  return sequence;
}


export {
  get_sbs_trinucleotide_contexts,
  standardize_substitution,
  init_sbs_mutational_spectra,
  standardize_trinucleotide,
  convertMatrix,
  getMutationalContext,
};
</code></pre>
        </article>
    </section>




</div>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Modules</h3><ul><li><a href="module-TCGA.html">TCGA</a></li><li><a href="module-User%252520Data.html">User Data</a></li><li><a href="module-Utils.html">Utils</a></li><li><a href="module-mSigPortal%2520APIs.html">mSigPortal APIs</a></li><li><a href="module-machineLearning.html">machineLearning</a></li></ul><h3>Global</h3><ul><li><a href="global.html#fitMutationalSpectraToSignatures">fitMutationalSpectraToSignatures</a></li></ul>
</nav>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc/jsdoc">JSDoc 4.0.4</a> on Fri Dec 27 2024 20:22:59 GMT-0500 (Eastern Standard Time)
</footer>

<script> prettyPrint(); </script>
<script src="scripts/linenumber.js"> </script>
</body>
</html>
