<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>JSDoc: Namespace: mSigPortalPlots</title>

    <script src="scripts/prettify/prettify.js"> </script>
    <script src="scripts/prettify/lang-css.js"> </script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify-tomorrow.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc-default.css">
</head>

<body>

<div id="main">

    <h1 class="page-title">Namespace: mSigPortalPlots</h1>

    




<section>

<header>
    
        <h2>mSigPortalPlots</h2>
        
    
</header>

<article>
    <div class="container-overview">
    
        

        


<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="main.js.html">main.js</a>, <a href="main.js.html#line86">line 86</a>
    </li></ul></dd>
    

    

    

    
</dl>


        
    
    </div>

    

    

    

    

    

    

    

    
        <h3 class="subsection-title">Methods</h3>

        
            

    

    
    <h4 class="name" id=".MSPrevalence"><span class="type-signature">(static) </span>MSPrevalence<span class="signature">(data, minimum)</span><span class="type-signature"> &rarr; {Object}</span></h4>
    

    



<div class="description">
    Calculates and visualizes the prevalence of mutational signatures across a set of samples, grouped by cancer type. The function generates two plots: a pie chart showing the overall prevalence of each signature based on total mutations (exposure) and a bar chart displaying the frequency of each signature across samples, considering a minimum exposure threshold.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>data</code></td>
            

            <td class="type">
            
                
<span class="param-type">object</span>


            
            </td>

            

            

            <td class="description last">An object representing the mutational signature prevalence data. The `data` object is expected to be an array of objects with the following structure:  `[{ signatureName: "SBS1", sample: "sample1", exposure: 10, burden: 5 }, { signatureName: "SBS5", sample: "sample1", exposure: 20, burden: 5 }, ... ]`  Where `signatureName` is the name of a mutational signature (string), `sample` is a sample identifier (string), `exposure` is a non-negative number representing the exposure of that signature in the sample (can be 0), and `burden` is a numeric value representing the mutational burden for the sample (must be a number).</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>minimum</code></td>
            

            <td class="type">
            
                
<span class="param-type">number</span>
|

<span class="param-type">null</span>
|

<span class="param-type">undefined</span>


            
            </td>

            

            

            <td class="description last">The minimum exposure value for a signature in a sample to be considered prevalent in that sample. Samples with exposure below this threshold are not counted in the frequency calculation for the bar chart. If `minimum` is `null` or `undefined`, it defaults to 100.  - `null` or `undefined`: Sets the minimum exposure to 100.  - Any positive number: Sets the minimum exposure to that number.</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="main.js.html">main.js</a>, <a href="main.js.html#line1597">line 1597</a>
    </li></ul></dd>
    

    

    

    
</dl>















<h5>Returns:</h5>

        
<div class="param-desc">
    - Returns an object containing the `traces` and `layout` for a Plotly plot.  - `traces`: An array of trace objects. If the maximum frequency of signatures (considering the `minimum` threshold) is less than 1%, the array contains only a single pie chart trace. Otherwise, it contains a pie chart trace followed by multiple bar chart traces (one for each signature).  - `layout`: An object defining the layout of the plot, including title annotations, axis settings, and overall appearance. It includes conditional logic to handle cases where no signature has a frequency greater than 1%.
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">Object</span>


    </dd>
</dl>

    





        
            

    

    
    <h4 class="name" id=".MsAssociation"><span class="type-signature">(static) </span>MsAssociation<span class="signature">(data, signatureName1, signatureName2, both<span class="signature-attributes">opt</span>)</span><span class="type-signature"> &rarr; {object}</span></h4>
    

    



<div class="description">
    Calculates the association between two mutational signatures across a set of samples. It computes the linear regression, Pearson correlation, and Spearman correlation between the log-transformed exposures of the two signatures. The results are used to generate a scatter plot with marginal histograms, visualizing the relationship between the signatures.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        
        <th>Attributes</th>
        

        
        <th>Default</th>
        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>data</code></td>
            

            <td class="type">
            
                
<span class="param-type">Array.&lt;object></span>


            
            </td>

            
                <td class="attributes">
                

                

                
                </td>
            

            
                <td class="default">
                
                </td>
            

            <td class="description last">An array of objects representing the exposure data for a set of samples. Each object in the array should have the following properties:  - `sample`: A string representing the sample ID.  - `signatureName`: A string representing the name of the mutational signature.  - `exposure`: A numeric value representing the exposure of the signature in the sample. This value can theoretically range from 0 to infinity, although in practice, values are often normalized.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>signatureName1</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                

                

                
                </td>
            

            
                <td class="default">
                
                </td>
            

            <td class="description last">The name of the first mutational signature. This should match the `signatureName` values in the `data` array.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>signatureName2</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                

                

                
                </td>
            

            
                <td class="default">
                
                </td>
            

            <td class="description last">The name of the second mutational signature. This should also match the `signatureName` values in the `data` array. It can be the same as `signatureName1`.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>both</code></td>
            

            <td class="type">
            
                
<span class="param-type">boolean</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    false
                
                </td>
            

            <td class="description last">A boolean flag indicating whether to filter the data to include only samples where both signatures have non-zero exposure.  - `true`: Only samples with non-zero exposure to both signatures are included in the analysis. If `signatureName1` and `signatureName2` are the same, then no filtering occurs.  - `false`: All samples are included in the analysis, regardless of whether they have non-zero exposure to both signatures.</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="main.js.html">main.js</a>, <a href="main.js.html#line1562">line 1562</a>
    </li></ul></dd>
    

    

    

    
</dl>















<h5>Returns:</h5>

        
<div class="param-desc">
    - Returns an object containing the traces and layout for a Plotly plot.  - `traces`: An array of trace objects to be used in a Plotly plot. This includes the main scatter plot trace, the linear regression line trace, and two marginal histogram traces.  - `layout`: An object containing the layout configuration for a Plotly plot, including title, axis labels, annotations, and other visual properties.
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">object</span>


    </dd>
</dl>

    





        
            

    

    
    <h4 class="name" id=".plotCosineSimilarityHeatMap"><span class="type-signature">(async, static) </span>plotCosineSimilarityHeatMap<span class="signature">(groupedData, studyName<span class="signature-attributes">opt</span>, genomeDataType<span class="signature-attributes">opt</span>, cancerType<span class="signature-attributes">opt</span>, divID<span class="signature-attributes">opt</span>, conductDoubleClustering<span class="signature-attributes">opt</span>, colorscale<span class="signature-attributes">opt</span>, showTable<span class="signature-attributes">opt</span>)</span><span class="type-signature"> &rarr; {Promise.&lt;Array.&lt;Array.&lt;number>>>}</span></h4>
    

    



<div class="description">
    Generates a cosine similarity heatmap based on mutational spectra data.This function processes grouped mutational data to compute cosine similarities,optionally performs double clustering to reorder the data, and then visualizesthe similarities using a Plotly heatmap. It also supports displaying a tablerepresentation of the cosine similarity matrix alongside the heatmap.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        
        <th>Attributes</th>
        

        
        <th>Default</th>
        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>groupedData</code></td>
            

            <td class="type">
            
                
<span class="param-type">Object</span>


            
            </td>

            
                <td class="attributes">
                

                

                
                </td>
            

            
                <td class="default">
                
                </td>
            

            <td class="description last">The input data object where keys represent sample names  and values are objects representing mutational spectra. The mutational spectra  should be represented as key-value pairs where keys are mutation types and values  are counts or frequencies.  Example:  ```  {    'Sample1': {'C>A': 10, 'C>G': 15, 'C>T': 20, ...},    'Sample2': {'C>A': 5, 'C>G': 8, 'C>T': 12, ...},    ...  }  ```  The range of keys (mutation types) should be consistent across all samples. The values  (counts or frequencies) can be integers or floats, and their range can vary based on the  underlying data, but typically they are non-negative.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>studyName</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    "PCAWG"
                
                </td>
            

            <td class="description last">The name of the study. This is used in the  title of the heatmap. Common values include study identifiers like "PCAWG",  "TCGA", or specific project names. The parameter should be a string and can  technically accept any string value, but it is intended to represent the name  of a study or dataset.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>genomeDataType</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    "WGS"
                
                </td>
            

            <td class="description last">The type of genomic data. This is also  used in the title of the heatmap. Expected values typically include abbreviations  for common genomic data types such as "WGS" (Whole Genome Sequencing), "WES"  (Whole Exome Sequencing), "RNA-Seq", etc. Similar to `studyName`, any string  is technically accepted, but the intended use is to describe the data type.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>cancerType</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    "Lung-AdenoCA"
                
                </td>
            

            <td class="description last">The type of cancer. This is included  in the title of the heatmap. Common values are standard cancer type  abbreviations or names, like "Lung-AdenoCA" (Lung Adenocarcinoma), "BRCA"  (Breast Invasive Carcinoma), etc. Any string value is accepted, but it should  represent a specific cancer type.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>divID</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    "cosineSimilarityHeatMap"
                
                </td>
            

            <td class="description last">The ID of the HTML element where  the heatmap will be rendered. This should be a valid HTML element ID. If an element  with this ID does not exist, one will be created and appended to the document body.  Any string is accepted, but it should correspond to a unique ID in the HTML  document to avoid conflicts.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>conductDoubleClustering</code></td>
            

            <td class="type">
            
                
<span class="param-type">boolean</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    true
                
                </td>
            

            <td class="description last">A flag indicating whether to  perform double clustering (hierarchical clustering on both rows and columns)  on the cosine similarity matrix. If `true`, the rows and columns of the heatmap  will be reordered based on the clustering. If `false`, the order of samples in  `groupedData` will be maintained. Boolean values `true` or `false` are expected.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>colorscale</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    "RdBu"
                
                </td>
            

            <td class="description last">The Plotly colorscale to use for the heatmap.  This can be any valid Plotly colorscale name (e.g., "Viridis", "Greys", "YlGnBu",  "RdBu"). Plotly provides a wide range of predefined colorscales. The chosen  colorscale will affect the visual representation of the similarity values. Any  string is accepted but it should correspond to a valid Plotly colorscale for  optimal results.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>showTable</code></td>
            

            <td class="type">
            
                
<span class="param-type">boolean</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    false
                
                </td>
            

            <td class="description last">A flag indicating whether to display a table  representation of the cosine similarity matrix alongside the heatmap. If `true`,  a table will be rendered next to the heatmap. If `false`, only the heatmap will  be displayed. Boolean values `true` or `false` are expected.</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="main.js.html">main.js</a>, <a href="main.js.html#line792">line 792</a>
    </li></ul></dd>
    

    

    

    
</dl>













<h5>Throws:</h5>

        

    <div class="param-desc">
    
        Will throw an error if the `cosineSimilarity` function or the  `plotGraphWithPlotlyAndMakeDataDownloadable` function throws an error.
    
    </div>


    


<h5>Returns:</h5>

        
<div class="param-desc">
    A Promise that resolves to the cosine similarity matrix.  The matrix is a two-dimensional array of numbers, where each number represents  the cosine similarity between two samples. The values range from 0 to 1, where 1  indicates perfect similarity and 0 indicates no similarity. The dimensions of  the matrix will be NxN, where N is the number of samples in `groupedData`.
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">Promise.&lt;Array.&lt;Array.&lt;number>>></span>


    </dd>
</dl>

    





        
            

    

    
    <h4 class="name" id=".plotDatasetMutationalSignaturesExposure"><span class="type-signature">(static) </span>plotDatasetMutationalSignaturesExposure<span class="signature">(exposureData, divID, relative<span class="signature-attributes">opt</span>, datasetName<span class="signature-attributes">opt</span>, doubleCluster<span class="signature-attributes">opt</span>, colorscale<span class="signature-attributes">opt</span>)</span><span class="type-signature"> &rarr; {object}</span></h4>
    

    



<div class="description">
    Generates a heatmap visualizing the exposure of multiple samples to a set of mutational signatures within a dataset. The function provides options for displaying relative or absolute exposure values and for performing double hierarchical clustering to reorder the rows and columns of the heatmap. It also allows customization of the color scale used to represent exposure values.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        
        <th>Attributes</th>
        

        
        <th>Default</th>
        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>exposureData</code></td>
            

            <td class="type">
            
                
<span class="param-type">object</span>


            
            </td>

            
                <td class="attributes">
                

                

                
                </td>
            

            
                <td class="default">
                
                </td>
            

            <td class="description last">An object containing the exposure data for a set of samples. The structure of `exposureData` is expected to be:  `{ sampleId1: { signatureName1: exposureValue1, signatureName2: exposureValue2, ..., rnorm: number }, sampleId2: { signatureName1: exposureValue3, signatureName2: exposureValue4, ..., rnorm: number }, ... }`  The outer keys (e.g., `sampleId1`, `sampleId2`) are sample identifiers (strings). The inner objects (e.g., `{ signatureName1: exposureValue1, ... }`) represent the exposure values for a given sample. `signatureName` keys are strings representing the names of mutational signatures (e.g., "SBS1", "SBS5"), and `exposureValue` are non-negative numbers representing the contribution of that signature to the sample. `rnorm` is a number that will be removed from the data before plotting.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>divID</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                

                

                
                </td>
            

            
                <td class="default">
                
                </td>
            

            <td class="description last">The ID of the HTML div element where the heatmap will be rendered.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>relative</code></td>
            

            <td class="type">
            
                
<span class="param-type">boolean</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    true
                
                </td>
            

            <td class="description last">A boolean indicating whether to display relative or absolute exposure values.  - `true`: The exposure values for each sample are normalized to sum to 1, representing the relative contribution of each signature.  - `false`: The raw exposure values are displayed.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>datasetName</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    "PCAWG"
                
                </td>
            

            <td class="description last">The name of the dataset being visualized. This is used as part of the plot title. Examples include:  - `"PCAWG"`  - `"TCGA"`  - Any other string that appropriately identifies the dataset.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>doubleCluster</code></td>
            

            <td class="type">
            
                
<span class="param-type">boolean</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    true
                
                </td>
            

            <td class="description last">A boolean indicating whether to perform double hierarchical clustering on the exposure data.  - `true`: The rows and columns of the heatmap are reordered based on the results of double clustering, which groups similar samples and signatures together.  - `false`: The rows and columns are displayed in the order they appear in the input `exposureData`.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>colorscale</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>
|

<span class="param-type">Array</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    "Custom"
                
                </td>
            

            <td class="description last">The color scale to use for the heatmap. Possible values are:  - `"Custom"`: A predefined custom color scale designed for visualizing exposure data.  - Any valid Plotly color scale name (e.g., `"Viridis"`, `"Blues"`, `"Hot"`, etc.).  - An array of arrays defining a custom color scale, where each inner array specifies a color stop with a value between 0 and 1 and a corresponding RGB color string (e.g., `[["0.0", "rgb(49,54,149)"], ["1.0", "rgb(165,0,38)"]]`).</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="main.js.html">main.js</a>, <a href="main.js.html#line1440">line 1440</a>
    </li></ul></dd>
    

    

    

    
</dl>















<h5>Returns:</h5>

        
<div class="param-desc">
    - Returns the data object used by Plotly to generate the heatmap. This object contains the z values (exposure values), x values (signature names), y values (sample names), and other settings for the heatmap, including the color scale. The structure is:  `{ z: [[exposureValue1, exposureValue2, ...], [exposureValue3, exposureValue4, ...], ...], x: [signatureName1, signatureName2, ...], y: [sampleId1, sampleId2, ...], type: "heatmap", colorscale: colorscale }`
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">object</span>


    </dd>
</dl>

    





        
            

    

    
    <h4 class="name" id=".plotForceDirectedTree"><span class="type-signature">(static) </span>plotForceDirectedTree<span class="signature">(groupedData, studyName<span class="signature-attributes">opt</span>, genomeDataType<span class="signature-attributes">opt</span>, cancerType<span class="signature-attributes">opt</span>, divID<span class="signature-attributes">opt</span>)</span><span class="type-signature"> &rarr; {object}</span></h4>
    

    



<div class="description">
    This function generates and displays a force-directed tree representing the relationships between patients in a study based on their mutational spectra. It calculates the cosine similarity between the mutational spectra of patients, performs hierarchical clustering based on these similarities, and then visualizes the resulting clusters as a force-directed tree.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        
        <th>Attributes</th>
        

        
        <th>Default</th>
        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>groupedData</code></td>
            

            <td class="type">
            
                
<span class="param-type">object</span>


            
            </td>

            
                <td class="attributes">
                

                

                
                </td>
            

            
                <td class="default">
                
                </td>
            

            <td class="description last">An object where keys represent sample IDs and values are objects containing mutational spectra data. The structure of `groupedData` is expected to be:  `{ sampleId1: { mutationType1: count1, mutationType2: count2, ... }, sampleId2: { mutationType1: count3, mutationType2: count4, ... }, ... }`  The inner objects (e.g., `{ mutationType1: count1, ... }`) represent the mutational spectrum for a given sample. `mutationType` keys can be any string representing a type of mutation (e.g., "C>A", "T>G"), and `count` values are non-negative integers representing the number of times that mutation type is observed in the sample.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>studyName</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    "PCAWG"
                
                </td>
            

            <td class="description last">The name of the study. This is used for labeling purposes in the visualization. Common values include, but are not limited to:  - `"PCAWG"`  - `"TCGA"`  - Any other string representing a specific study.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>genomeDataType</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    "WGS"
                
                </td>
            

            <td class="description last">The type of genome data used. This is also used for labeling purposes. Possible values include:  - `"WGS"`: Whole Genome Sequencing  - `"WES"`: Whole Exome Sequencing  - `"RNA-Seq"`: RNA Sequencing</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>cancerType</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    "Lung-AdenoCA"
                
                </td>
            

            <td class="description last">The type of cancer being studied. This is used for labeling in the visualization. Examples include:  - `"Lung-AdenoCA"`: Lung Adenocarcinoma  - `"Breast-AdenoCA"`: Breast Adenocarcinoma  - Any valid cancer type identifier.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>divID</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    "forceDirectedTree"
                
                </td>
            

            <td class="description last">The ID of the HTML div element where the force-directed tree will be rendered.</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="main.js.html">main.js</a>, <a href="main.js.html#line1056">line 1056</a>
    </li></ul></dd>
    

    

    

    
</dl>















<h5>Returns:</h5>

        
<div class="param-desc">
    - Returns the formatted hierarchical clusters used to generate the force-directed tree. The structure of this object is compatible with the AM5 charting library and represents the hierarchical relationships between samples based on their mutational spectra. The format is a nested object where each level represents a node in the tree. Each node can have properties such as `name`, `value`, `children` (an array of child nodes), and potentially others added during formatting.
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">object</span>


    </dd>
</dl>

    





        
            

    

    
    <h4 class="name" id=".plotMSPrevalenceData"><span class="type-signature">(static) </span>plotMSPrevalenceData<span class="signature">(divID, data)</span><span class="type-signature"> &rarr; {void}</span></h4>
    

    



<div class="description">
    This function is a wrapper around the `plotMSPrevalence` function. It takes the output of `plotMSPrevalence` and uses it to generate a Plotly plot, which is then displayed in a specified div. The plot visualizes the prevalence of mutational signatures.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>divID</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">The ID of the div element where the plot will be rendered.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>data</code></td>
            

            <td class="type">
            
                
<span class="param-type">object</span>


            
            </td>

            

            

            <td class="description last">An object representing the mutational signature prevalence data. The `data` object is expected to be an array of objects with the following structure:  `[{ signatureName: "SBS1", sample: "sample1", exposure: 10, burden: 5 }, { signatureName: "SBS5", sample: "sample1", exposure: 20, burden: 5 }, ... ]`  Where `signatureName` is the name of a mutational signature (string), `sample` is a sample identifier (string), `exposure` is a non-negative number representing the exposure of that signature in the sample, and `burden` is a numeric value representing the mutational burden for the sample.</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="main.js.html">main.js</a>, <a href="main.js.html#line1586">line 1586</a>
    </li></ul></dd>
    

    

    

    
</dl>















<h5>Returns:</h5>

        
<div class="param-desc">
    - This function does not return a value. It directly renders the plot in the specified `divID`.
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">void</span>


    </dd>
</dl>

    





        
            

    

    
    <h4 class="name" id=".plotPatientMutationalSignaturesExposure"><span class="type-signature">(static) </span>plotPatientMutationalSignaturesExposure<span class="signature">(exposureData, divID, sample)</span><span class="type-signature"> &rarr; {object}</span></h4>
    

    



<div class="description">
    Generates a pie chart visualizing the exposure of a single sample to a set of mutational signatures. The function takes exposure data, which includes the relative contribution of each signature to the sample's mutational profile, and displays it in a pie chart format.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>exposureData</code></td>
            

            <td class="type">
            
                
<span class="param-type">object</span>


            
            </td>

            

            

            <td class="description last">An object containing the exposure data for a set of samples. The structure of `exposureData` is expected to be:  `{ sampleId1: { signatureName1: exposureValue1, signatureName2: exposureValue2, ... }, sampleId2: { signatureName1: exposureValue3, signatureName2: exposureValue4, ... }, ... }`  The outer keys (e.g., `sampleId1`, `sampleId2`) are sample identifiers (strings). The inner objects (e.g., `{ signatureName1: exposureValue1, ... }`) represent the exposure values for a given sample. `signatureName` keys are strings representing the names of mutational signatures (e.g., "SBS1", "SBS5"), and `exposureValue` are non-negative numbers representing the contribution of that signature to the sample. These values typically sum to 1 for each sample. The `exposureData` object can contain multiple samples, but only the data for the specified `sample` will be used for plotting. `exposureData` must also have a `rnorm` property which is a number.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>divID</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">The ID of the HTML div element where the pie chart will be rendered.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>sample</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">The ID of the sample for which to plot the mutational signature exposure. This should be one of the keys in the `exposureData` object (e.g., "sampleId1", "sampleId2").</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="main.js.html">main.js</a>, <a href="main.js.html#line1395">line 1395</a>
    </li></ul></dd>
    

    

    

    
</dl>















<h5>Returns:</h5>

        
<div class="param-desc">
    - Returns the data object used by Plotly to generate the pie chart. This object contains the labels (signature names), values (exposure values), and other settings for the pie chart. The format is:  `{ labels: [signatureName1, signatureName2, ...], values: [exposureValue1, exposureValue2, ...], name: "sample exposure values", textposition: "inside", hole: 0.4, hoverinfo: "name + value", type: "pie" }`
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">object</span>


    </dd>
</dl>

    





        
            

    

    
    <h4 class="name" id=".plotPatientMutationalSpectrum"><span class="type-signature">(async, static) </span>plotPatientMutationalSpectrum<span class="signature">(mutationalSpectra, matrixSize<span class="signature-attributes">opt</span>, divID<span class="signature-attributes">opt</span>)</span><span class="type-signature"> &rarr; {Promise.&lt;void>}</span></h4>
    

    



<div class="description">
    Renders a plot of the mutational spectra for one or more patients in a given div element ID using Plotly.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        
        <th>Attributes</th>
        

        
        <th>Default</th>
        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>mutationalSpectra</code></td>
            

            <td class="type">
            
                
<span class="param-type">Object</span>


            
            </td>

            
                <td class="attributes">
                

                

                
                </td>
            

            
                <td class="default">
                
                </td>
            

            <td class="description last">An object containing the mutational spectra data for one or more patients.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>matrixSize</code></td>
            

            <td class="type">
            
                
<span class="param-type">number</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    96
                
                </td>
            

            <td class="description last">The size of the plot matrix. Defaults to 96.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>divID</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    'mutationalSpectrumMatrix'
                
                </td>
            

            <td class="description last">The ID of the div element to render the plot in. Defaults to 'mutationalSpectrumMatrix'.</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="main.js.html">main.js</a>, <a href="main.js.html#line410">line 410</a>
    </li></ul></dd>
    

    

    

    
</dl>













<h5>Throws:</h5>

        

<dl>
    <dt>
        <div class="param-desc">
        An error is thrown if no data is available for the selected parameters.
        </div>
    </dt>
    <dd></dd>
    <dt>
        <dl>
            <dt>
                Type
            </dt>
            <dd>
                
<span class="param-type">Error</span>


            </dd>
        </dl>
    </dt>
    <dd></dd>
</dl>


    


<h5>Returns:</h5>

        
<div class="param-desc">
    A promise that resolves when the plot has been rendered.
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">Promise.&lt;void></span>


    </dd>
</dl>

    





        
            

    

    
    <h4 class="name" id=".plotPatientMutationalSpectrumuserData"><span class="type-signature">(async, static) </span>plotPatientMutationalSpectrumuserData<span class="signature">(mutationalSpectra, matrixSize<span class="signature-attributes">opt</span>, divID<span class="signature-attributes">opt</span>)</span><span class="type-signature"></span></h4>
    

    



<div class="description">
    Plots the mutational spectrum for the given parameters.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        
        <th>Attributes</th>
        

        
        <th>Default</th>
        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>mutationalSpectra</code></td>
            

            <td class="type">
            
                
<span class="param-type">Object</span>


            
            </td>

            
                <td class="attributes">
                

                

                
                </td>
            

            
                <td class="default">
                
                </td>
            

            <td class="description last">An object containing the mutational spectra data.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>matrixSize</code></td>
            

            <td class="type">
            
                
<span class="param-type">number</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    96
                
                </td>
            

            <td class="description last">The size of the matrix to be plotted.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>divID</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    "mutationalSpectrumMatrix"
                
                </td>
            

            <td class="description last">The ID of the div element where the plot will be displayed.</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="main.js.html">main.js</a>, <a href="main.js.html#line348">line 348</a>
    </li></ul></dd>
    

    

    

    
</dl>




















        
            

    

    
    <h4 class="name" id=".plotProfilerSummary"><span class="type-signature">(async, static) </span>plotProfilerSummary<span class="signature">(studyName<span class="signature-attributes">opt</span>, genomeDataType<span class="signature-attributes">opt</span>, cancerTypeOrGroup<span class="signature-attributes">opt</span>, numberOfResults<span class="signature-attributes">opt</span>, divID<span class="signature-attributes">opt</span>)</span><span class="type-signature"> &rarr; {Promise.&lt;void>}</span></h4>
    

    



<div class="description">
    Generates a mutational spectrum summary plot and displays it in a given HTML div element.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        
        <th>Attributes</th>
        

        
        <th>Default</th>
        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>studyName</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    "PCAWG"
                
                </td>
            

            <td class="description last">The name of the cancer genomics study to use. Default is "PCAWG".</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>genomeDataType</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    "WGS"
                
                </td>
            

            <td class="description last">The type of genomic data to use. Default is "WGS".</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>cancerTypeOrGroup</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    "Lung-AdenoCA"
                
                </td>
            

            <td class="description last">The cancer type or group to display. Default is "Lung-AdenoCA".</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>numberOfResults</code></td>
            

            <td class="type">
            
                
<span class="param-type">number</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    50
                
                </td>
            

            <td class="description last">The maximum number of results to display. Default is 50.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>divID</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    "mutationalSpectrumSummary"
                
                </td>
            

            <td class="description last">The ID of the HTML div element where the plot will be displayed. Default is "mutationalSpectrumSummary".</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="main.js.html">main.js</a>, <a href="main.js.html#line179">line 179</a>
    </li></ul></dd>
    

    

    

    
</dl>













<h5>Throws:</h5>

        

<dl>
    <dt>
        <div class="param-desc">
        If there is an error retrieving or displaying the plot, this function will throw an Error with a message describing the error.
        </div>
    </dt>
    <dd></dd>
    <dt>
        <dl>
            <dt>
                Type
            </dt>
            <dd>
                
<span class="param-type">Error</span>


            </dd>
        </dl>
    </dt>
    <dd></dd>
</dl>


    


<h5>Returns:</h5>

        
<div class="param-desc">
    A Promise that resolves when the plot is displayed or rejects if there is an error.
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">Promise.&lt;void></span>


    </dd>
</dl>

    





        
            

    

    
    <h4 class="name" id=".plotProjectMutationalBurdenByCancerType"><span class="type-signature">(async, static) </span>plotProjectMutationalBurdenByCancerType<span class="signature">(project, divID)</span><span class="type-signature"> &rarr; {Promise}</span></h4>
    

    



<div class="description">
    Plots the mutational burden by cancer type for a given project.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>project</code></td>
            

            <td class="type">
            
                
<span class="param-type">Object</span>


            
            </td>

            

            

            <td class="description last">An object containing mutational data for different cancer types.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>divID</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">The ID of the div where the plot should be displayed.</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="main.js.html">main.js</a>, <a href="main.js.html#line271">line 271</a>
    </li></ul></dd>
    

    

    

    
</dl>















<h5>Returns:</h5>

        
<div class="param-desc">
    - A Promise that resolves when the plot is displayed.
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">Promise</span>


    </dd>
</dl>

    




    <h5>Example</h5>
    
    <pre class="prettyprint"><code>// Example usage:
plotProjectMutationalBurdenByCancerType(projectData, "plotDiv");</code></pre>



        
            

    

    
    <h4 class="name" id=".plotSignatureActivityDataBy"><span class="type-signature">(static) </span>plotSignatureActivityDataBy<span class="signature">(divID, data, group<span class="signature-attributes">opt</span>)</span><span class="type-signature"> &rarr; {void}</span></h4>
    

    



<div class="description">
    Generates a box plot of signature activity data, grouped by a specified attribute. The function takes a dataset and groups it by the provided attribute (e.g., "signatureName", "study", "cancerType"). For each group, it creates a box trace where the y-values represent the log10 of the exposure values and the x-values are set to the group name. The box plot displays the distribution of exposure values for each group, with the option to show all individual data points (jittered for better visibility). Hovering over the data points reveals the sample name and the log10 of the exposure value. The plot also indicates the fraction of samples within each group that have non-zero exposure.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        
        <th>Attributes</th>
        

        
        <th>Default</th>
        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>divID</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                

                

                
                </td>
            

            
                <td class="default">
                
                </td>
            

            <td class="description last">The ID of the div element where the plot will be rendered.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>data</code></td>
            

            <td class="type">
            
                
<span class="param-type">Array.&lt;object></span>


            
            </td>

            
                <td class="attributes">
                

                

                
                </td>
            

            
                <td class="default">
                
                </td>
            

            <td class="description last">An array of objects representing the signature activity data. Each object must have at least the following properties:  - `exposure`: A numeric value representing the exposure of a signature. It can be any positive number or zero.  - `sample`: A string representing the sample ID.  - The `data` array must also contain a property matching the name specified by the `group` parameter (e.g., "signatureName", "study", "cancerType").</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>group</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    "signatureName"
                
                </td>
            

            <td class="description last">The attribute to group the data by. Possible values are any property name present in the data objects, including, but not limited to:  - `"signatureName"`: Groups the data by signature names.  - `"study"`: Groups the data by study IDs.  - `"cancerType"`: Groups the data by cancer types.  - `"sample"`: Groups the data by sample IDs (Note: this might not result in a meaningful box plot).  - Any other custom property that exists in the data objects.  The default value is `"signatureName"`.</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="main.js.html">main.js</a>, <a href="main.js.html#line1002">line 1002</a>
    </li></ul></dd>
    

    

    

    
</dl>















<h5>Returns:</h5>

        
<div class="param-desc">
    - This function does not return a value. It directly renders the plot in the specified `divID`.
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">void</span>


    </dd>
</dl>

    





        
            

    

    
    <h4 class="name" id=".plotSignatureAssociations"><span class="type-signature">(static) </span>plotSignatureAssociations<span class="signature">(divID, data, signature1, signature2)</span><span class="type-signature"> &rarr; {void}</span></h4>
    

    



<div class="description">
    This function generates and plots a scatter plot with marginal histograms, along with statistical analysis, to visualize the association between two mutational signatures. It calculates and displays the linear regression line, Pearson correlation, and Spearman correlation, providing insights into the relationship between the exposures of two signatures in a set of samples.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>divID</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">The ID of the HTML div element where the plot will be rendered.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>data</code></td>
            

            <td class="type">
            
                
<span class="param-type">object</span>


            
            </td>

            

            

            <td class="description last">An array of objects representing the exposure data for a set of samples. Each object in the array should have the following properties:  - `sample`: A string representing the sample ID.  - `signatureName`: A string representing the name of the mutational signature.  - `exposure`: A numeric value representing the exposure of the signature in the sample.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>signature1</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">The name of the first mutational signature. This should match the `signatureName` values in the `data` array. The values can be any valid signature name present in the dataset, for example, "SBS1", "SBS5", "DBS1", "ID4".</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>signature2</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            

            

            <td class="description last">The name of the second mutational signature. This should also match the `signatureName` values in the `data` array. Similar to `signature1`, the values can be any signature name present in the dataset and can also be the same as `signature1` to assess the distribution of a single signature.</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="main.js.html">main.js</a>, <a href="main.js.html#line1548">line 1548</a>
    </li></ul></dd>
    

    

    

    
</dl>















<h5>Returns:</h5>

        
<div class="param-desc">
    - This function does not return a value. It directly renders the plot in the specified `divID`.
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">void</span>


    </dd>
</dl>

    





        
            

    

    
    <h4 class="name" id=".plotUMAPVisualization"><span class="type-signature">(static) </span>plotUMAPVisualization<span class="signature">(data, datasetName<span class="signature-attributes">opt</span>, divID, nComponents<span class="signature-attributes">opt</span>, minDist<span class="signature-attributes">opt</span>, nNeighbors<span class="signature-attributes">opt</span>)</span><span class="type-signature"> &rarr; {object}</span></h4>
    

    



<div class="description">
    Generates a UMAP (Uniform Manifold Approximation and Projection) visualization of mutational spectra data. UMAP is a dimensionality reduction technique used to project high-dimensional data into a lower-dimensional space (typically 2D or 3D) while preserving the global structure of the data. This function takes mutational spectra data, applies UMAP to reduce its dimensionality, and then creates either a 2D or 3D scatter plot to visualize the results. If `nComponents` is set to 3, it additionally generates a mesh3d trace to highlight the density of points in the 3D space.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        
        <th>Attributes</th>
        

        
        <th>Default</th>
        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>data</code></td>
            

            <td class="type">
            
                
<span class="param-type">object</span>


            
            </td>

            
                <td class="attributes">
                

                

                
                </td>
            

            
                <td class="default">
                
                </td>
            

            <td class="description last">An object representing the mutational spectra data. The structure of `data` is expected to be:  `{ sampleId1: { mutationType1: count1, mutationType2: count2, ... }, sampleId2: { mutationType1: count3, mutationType2: count4, ... }, ... }`  The outer keys (e.g., `sampleId1`, `sampleId2`) are sample identifiers (strings). The inner objects (e.g., `{ mutationType1: count1, ... }`) represent the mutational spectrum for a given sample. `mutationType` keys can be any string representing a type of mutation (e.g., "C>A", "T>G"), and `count` values are non-negative integers representing the number of times that mutation type is observed in the sample.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>datasetName</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    "PCAWG"
                
                </td>
            

            <td class="description last">The name of the dataset being visualized. This is used as part of the plot title. Examples include:  - `"PCAWG"`  - `"TCGA"`  - Any other string that appropriately identifies the dataset.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>divID</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                

                

                
                </td>
            

            
                <td class="default">
                
                </td>
            

            <td class="description last">The ID of the HTML div element where the plot will be rendered.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>nComponents</code></td>
            

            <td class="type">
            
                
<span class="param-type">number</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    3
                
                </td>
            

            <td class="description last">The number of dimensions to reduce the data to using UMAP. This determines whether a 2D or 3D plot is generated. Possible values are:  - `2`: Generates a 2D scatter plot.  - `3`: Generates a 3D scatter plot with an additional mesh3d trace.  Any other positive integer is technically permissible but may not yield meaningful visualizations.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>minDist</code></td>
            

            <td class="type">
            
                
<span class="param-type">number</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    0.1
                
                </td>
            

            <td class="description last">The effective minimum distance between embedded points in the UMAP projection. Smaller values result in a more clustered embedding, while larger values preserve more of the global structure. The valid range is between 0.0 and 1.0.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>nNeighbors</code></td>
            

            <td class="type">
            
                
<span class="param-type">number</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    15
                
                </td>
            

            <td class="description last">The number of neighboring points to consider when constructing the UMAP. Larger values capture more global structure in the data, while smaller values preserve more local structure. Values should be positive integers, typically in the range of 2 to 100.</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="main.js.html">main.js</a>, <a href="main.js.html#line1171">line 1171</a>
    </li></ul></dd>
    

    

    

    
</dl>















<h5>Returns:</h5>

        
<div class="param-desc">
    - Returns the trace object used by Plotly to generate the visualization. This object contains the data points, plot type, marker settings, and, in the case of a 3D plot, the mesh3d settings. The structure depends on the value of `nComponents`.
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">object</span>


    </dd>
</dl>

    





        
    

    

    
</article>

</section>




</div>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Namespaces</h3><ul><li><a href="mSigPortalData.html">mSigPortalData</a></li><li><a href="mSigPortalPlots.html">mSigPortalPlots</a></li><li><a href="machineLearning.html">machineLearning</a></li><li><a href="tcga.html">tcga</a></li><li><a href="userData.html">userData</a></li></ul>
</nav>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc/jsdoc">JSDoc 4.0.4</a> on Thu Jan 02 2025 18:22:29 GMT-0500 (Eastern Standard Time)
</footer>

<script> prettyPrint(); </script>
<script src="scripts/linenumber.js"> </script>
</body>
</html>