<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>JSDoc: Namespace: userData</title>

    <script src="scripts/prettify/prettify.js"> </script>
    <script src="scripts/prettify/lang-css.js"> </script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify-tomorrow.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc-default.css">
</head>

<body>

<div id="main">

    <h1 class="page-title">Namespace: userData</h1>

    




<section>

<header>
    
        <h2>userData</h2>
        
    
</header>

<article>
    <div class="container-overview">
    
        

        


<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="main.js.html">main.js</a>, <a href="main.js.html#line95">line 95</a>
    </li></ul></dd>
    

    

    

    
</dl>


        
    
    </div>

    

    

    

    

    

    

    

    
        <h3 class="subsection-title">Methods</h3>

        
            

    

    
    <h4 class="name" id=".convertMatrix"><span class="type-signature">(async, static) </span>convertMatrix<span class="signature">(data, group_by<span class="signature-attributes">opt</span>, batch_size<span class="signature-attributes">opt</span>, genome<span class="signature-attributes">opt</span>, tcga<span class="signature-attributes">opt</span>)</span><span class="type-signature"> &rarr; {Promise.&lt;Object>}</span></h4>
    

    



<div class="description">
    Converts input mutational data into a mutational spectrum matrix grouped by a specified field.This function processes raw mutational data, extracts trinucleotide contexts, and aggregates mutational spectra for each group. It supports TCGA and non-TCGA formats, allowing batch processing for large datasets. The resulting mutational spectrum matrix is formatted for downstream visualization and analysis.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        
        <th>Attributes</th>
        

        
        <th>Default</th>
        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>data</code></td>
            

            <td class="type">
            
                
<span class="param-type">Array.&lt;Array.&lt;Object>></span>


            
            </td>

            
                <td class="attributes">
                

                

                
                </td>
            

            
                <td class="default">
                
                </td>
            

            <td class="description last">Array of patient-level mutational data. Each patient's data is represented as an array of objects, where each object contains mutational details (e.g., chromosome, position, mutation type).</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>group_by</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    "Center"
                
                </td>
            

            <td class="description last">Field to group data by (e.g., "Center" or "sample_id"). This field should exist in the input data.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>batch_size</code></td>
            

            <td class="type">
            
                
<span class="param-type">number</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    100
                
                </td>
            

            <td class="description last">Number of mutations to process in parallel batches. Adjust this for memory management.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>genome</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    "hg19"
                
                </td>
            

            <td class="description last">Reference genome build. Defaults to "hg19" unless specified in the data.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>tcga</code></td>
            

            <td class="type">
            
                
<span class="param-type">boolean</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    false
                
                </td>
            

            <td class="description last">Flag indicating whether the input data is in TCGA format. If true, expects TCGA-specific fields.</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="mSigSDKScripts_mutationalSpectrum.js.html">mSigSDKScripts/mutationalSpectrum.js</a>, <a href="mSigSDKScripts_mutationalSpectrum.js.html#line102">line 102</a>
    </li></ul></dd>
    

    

    

    
</dl>















<h5>Returns:</h5>

        
<div class="param-desc">
    - A promise resolving to an object where each key is a group (e.g., "CNIC"),and each value is a mutational spectrum object. The mutational spectrum object contains trinucleotide contexts as keys (e.g., "A[C>A]A") and counts as values.
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">Promise.&lt;Object></span>


    </dd>
</dl>

    




    <h5>Example</h5>
    
    <pre class="prettyprint"><code>// Example input dataconst data = [  [    { chromosome: "1", start_position: "12345", reference_allele: "C", tumor_seq_allele2: "T", variant_type: "SNP", build: "hg19", Center: "CNIC" },    { chromosome: "2", start_position: "67890", reference_allele: "G", tumor_seq_allele2: "A", variant_type: "SNP", build: "hg19", Center: "CNIC" }  ],  [    { chromosome: "3", start_position: "101112", reference_allele: "T", tumor_seq_allele2: "C", variant_type: "SNP", build: "hg19", Center: "OtherCenter" }  ]];// Convert data to mutational spectra grouped by Centerconst mutationalSpectra = await convertMatrix(data, "Center", 50, "hg19", false);console.log(mutationalSpectra);// Output:// {//   "CNIC": {//     "A[C>A]A": 9,//     "A[C>A]C": 7,//     "A[C>A]G": 6,//     ...//   },//   "OtherCenter": {//     "T[T>C]A": 15,//     "T[T>C]T": 8,//     ...//   }// }</code></pre>



        
            

    

    
    <h4 class="name" id=".convertMutationalSpectraIntoJSON"><span class="type-signature">(static) </span>convertMutationalSpectraIntoJSON<span class="signature">(MAFfiles, mutSpec, sample_name, dataType<span class="signature-attributes">opt</span>)</span><span class="type-signature"> &rarr; {Array}</span></h4>
    

    



<div class="description">
    Converts mutational spectra into JSON objects for downstream processing or storage.This function takes mutation annotation format (MAF) files, mutational spectra data, and sample names, and outputs JSON objects representing the spectra in a structured format.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        
        <th>Attributes</th>
        

        
        <th>Default</th>
        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>MAFfiles</code></td>
            

            <td class="type">
            
                
<span class="param-type">Array</span>


            
            </td>

            
                <td class="attributes">
                

                

                
                </td>
            

            
                <td class="default">
                
                </td>
            

            <td class="description last">An array of arrays containing mutation annotation data for each sample. Each inner array represents a sample's mutation data, where each entry is an object with key-value pairs.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>mutSpec</code></td>
            

            <td class="type">
            
                
<span class="param-type">Object</span>


            
            </td>

            
                <td class="attributes">
                

                

                
                </td>
            

            
                <td class="default">
                
                </td>
            

            <td class="description last">An object representing the mutational spectra. Keys are patient identifiers, and values are objects with mutation types as keys and counts as values.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>sample_name</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                

                

                
                </td>
            

            
                <td class="default">
                
                </td>
            

            <td class="description last">The key in MAFfiles to be used as the sample identifier.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>dataType</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>


            
            </td>

            
                <td class="attributes">
                
                    &lt;optional><br>
                

                

                
                </td>
            

            
                <td class="default">
                
                    "WGS"
                
                </td>
            

            <td class="description last">The sequencing strategy, e.g., "WGS" (Whole Genome Sequencing) or "WES" (Whole Exome Sequencing).</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="mSigSDKScripts_userData.js.html">mSigSDKScripts/userData.js</a>, <a href="mSigSDKScripts_userData.js.html#line119">line 119</a>
    </li></ul></dd>
    

    

    

    
</dl>













<h5>Throws:</h5>

        

<dl>
    <dt>
        <div class="param-desc">
        - If the number of MAF files and the number of mutational spectra do not match.
        </div>
    </dt>
    <dd></dd>
    <dt>
        <dl>
            <dt>
                Type
            </dt>
            <dd>
                
<span class="param-type">Error</span>


            </dd>
        </dl>
    </dt>
    <dd></dd>
</dl>


    


<h5>Returns:</h5>

        
<div class="param-desc">
    - An array of arrays, where each inner array represents the JSON objects for a patient's mutational spectra.
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">Array</span>


    </dd>
</dl>

    




    <h5>Examples</h5>
    
    <pre class="prettyprint"><code>// Example input:const MAFfiles = [  [{ sample_id: "Sample1", ch   mosome: "1", start_position: "12345", ... }],  [{ sample_id: "Sample2", ch   mosome: "2", start_position: "67890", ... }]];const mutSpec = {  Patient1: { "A[C>A]A": 10,    [C>A]C": 5, "A[C>A]G": 8, ... },  Patient2: { "A[C>A]A": 7, "   C>A]C": 4, "A[C>A]G": 6, ... }};const sample_name = "sample_id";const result = convertMutationalSpectraIntoJSON(MAFfiles, mutSpec, sample_name, "WES");console.log(result);</code></pre>

    <pre class="prettyprint"><code>// Example output:[  [    { sample: "Sample1", strategy: "WES", profile: "SBS", matrix: 96, mutationType: "C>A", mutations: 10 },    { sample: "Sample1", strategy: "WES", profile: "SBS", matrix: 96, mutationType: "C>G", mutations: 5 },    ...  ],  [    { sample: "Sample2", strategy: "WES", profile: "SBS", matrix: 96, mutationType: "C>A", mutations: 7 },    { sample: "Sample2", strategy: "WES", profile: "SBS", matrix: 96, mutationType: "C>G", mutations: 4 },    ...  ]];</code></pre>



        
            

    

    
    <h4 class="name" id=".convertWGStoPanel"><span class="type-signature">(async, static) </span>convertWGStoPanel<span class="signature">(WgMAFs, panelDf)</span><span class="type-signature"> &rarr; {Array}</span></h4>
    

    



<div class="description">
    Converts Whole Genome Sequencing (WGS) mutation data into panel data by downsampling based on a BED file.This function takes WGS mutation annotation files (MAFs) and a BED file defining panel regions, then filtersthe WGS data to include only mutations within the specified panel regions.
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>WgMAFs</code></td>
            

            <td class="type">
            
                
<span class="param-type">Array</span>


            
            </td>

            

            

            <td class="description last">An array of WGS mutation data, where each element is an array representing mutations for a single sample.Each mutation record should be an object with fields such as `chromosome`, `start_position`, etc.</td>
        </tr>

    

        <tr>
            
                <td class="name"><code>panelDf</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>
|

<span class="param-type">Array</span>


            
            </td>

            

            

            <td class="description last">A BED file defining the regions of the panel or an array of arrays representing the panel regions.If a string is provided, it is treated as a file path to a BED file and read into memory.</td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="mSigSDKScripts_userData.js.html">mSigSDKScripts/userData.js</a>, <a href="mSigSDKScripts_userData.js.html#line53">line 53</a>
    </li></ul></dd>
    

    

    

    
</dl>















<h5>Returns:</h5>

        
<div class="param-desc">
    - An array of downsampled MAFs, where each element corresponds to a sample from the input WGS data,filtered to include only mutations within the panel regions.
</div>



<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">Array</span>


    </dd>
</dl>

    




    <h5>Example</h5>
    
    <pre class="prettyprint"><code>// Example input:const WgMAFs = [  [    { chromosome: "1", start_position: 12345, Hugo_Symbol: "TP53", ... },    { chromosome: "2", start_position: 67890, Hugo_Symbol: "BRCA1", ... },    ...  ],  [    { chromosome: "1", start_position: 54321, Hugo_Symbol: "KRAS", ... },    { chromosome: "2", start_position: 98765, Hugo_Symbol: "EGFR", ... },    ...  ],];const panelDf = "panel_regions.bed"; // Path to a BED file.// Convert WGS data to panel dataconst panelMAFs = await convertWGStoPanel(WgMAFs, panelDf);// Example output:// panelMAFs = [//   [//     { chromosome: "1", start_position: 12345, Hugo_Symbol: "TP53", ... },//     ...//   ],//   [//     { chromosome: "2", start_position: 98765, Hugo_Symbol: "EGFR", ... },//     ...//   ]// ];</code></pre>



        
    

    

    
</article>

</section>




</div>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Namespaces</h3><ul><li><a href="mSigPortalData.html">mSigPortalData</a></li><li><a href="mSigPortalPlots.html">mSigPortalPlots</a></li><li><a href="machineLearning.html">machineLearning</a></li><li><a href="tcga.html">tcga</a></li><li><a href="userData.html">userData</a></li></ul>
</nav>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc/jsdoc">JSDoc 4.0.4</a> on Sun Dec 29 2024 18:18:19 GMT-0500 (Eastern Standard Time)
</footer>

<script> prettyPrint(); </script>
<script src="scripts/linenumber.js"> </script>
</body>
</html>